source/loader/layers/sanitizer/asan_interceptor.hpp

```diff
@@ -0,0 +1,143 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file ur_sanitizer_layer.cpp
+ *
+ */
+
+#pragma once
+
+#include "common.hpp"
+
+#include <map>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+namespace ur_sanitizer_layer {
+
+enum USMMemoryType { DEVICE, SHARE, HOST, MEM_BUFFER };
+
+struct USMAllocInfo {
+    uptr AllocBegin;
+    uptr UserBegin;
+    uptr UserEnd;
+    size_t AllocSize;
+    USMMemoryType Type;
+};
+
+enum class DeviceType { UNKNOWN, CPU, GPU_PVC, GPU_DG2 };
+
+struct DeviceInfo {
+    DeviceType Type;
+    size_t Alignment;
+    uptr ShadowOffset;
+    uptr ShadowOffsetEnd;
+
+    // Lock InitPool & AllocInfos
+    ur_shared_mutex Mutex;
+    std::vector<std::shared_ptr<USMAllocInfo>> AllocInfos;
+};
+
+struct QueueInfo {
+    ur_mutex Mutex;
+    ur_event_handle_t LastEvent;
+};
+
+struct ContextInfo {
+
+    std::shared_ptr<DeviceInfo> getDeviceInfo(ur_device_handle_t Device) {
+        std::shared_lock<ur_shared_mutex> Guard(Mutex);
+        assert(DeviceMap.find(Device) != DeviceMap.end());
+        return DeviceMap[Device];
+    }
+
+    std::shared_ptr<QueueInfo> getQueueInfo(ur_queue_handle_t Queue) {
+        std::shared_lock<ur_shared_mutex> Guard(Mutex);
+        assert(QueueMap.find(Queue) != QueueMap.end());
+        return QueueMap[Queue];
+    }
+
+    std::shared_ptr<USMAllocInfo> getUSMAllocInfo(uptr Address) {
+        std::shared_lock<ur_shared_mutex> Guard(Mutex);
+        assert(AllocatedUSMMap.find(Address) != AllocatedUSMMap.end());
+        return AllocatedUSMMap[Address];
+    }
+
+    ur_shared_mutex Mutex;
+    // Note: nullptr is host device
+    std::unordered_map<ur_device_handle_t, std::shared_ptr<DeviceInfo>>
+        DeviceMap;
+    std::unordered_map<ur_queue_handle_t, std::shared_ptr<QueueInfo>> QueueMap;
+
+    /// key: USMAllocInfo.AllocBegin
+    /// value: USMAllocInfo
+    /// Use AllocBegin as key can help to detect underflow pointer
+    std::map<uptr, std::shared_ptr<USMAllocInfo>> AllocatedUSMMap;
```

Comment:
it might be simpler to use an `unordered_map` indexed by the user pointer.

Reply:
This is intentional. I need to compare the value of pointer address.

Reply:
For the release logic I'm assuming? The only current use of this map is in `releaseMemory` where you are searching for the base allocation based on a user pointer. That can be done directly with a hashmap.